### ISSUE TITLE
Clone of volume gets swapped with primary

### ISSUE THREAD

**Issue by dvickernasa** [openzfs/zfs#15305]
### System information
Type | Version/Name
 --- | ---
Distribution Name	| CentOS
Distribution Version	| 7.9
Kernel Version	| 3.10.0-1160.95.1.el7.x86_64
Architecture	| x86_64
OpenZFS Version	| zfs-kmod-2.0.7-1

### Describe the problem you're observing

We have experienced situations where the primary dataset and a clone/snapshot of a ZFS volume get swapped.  To the high level ZFS commands, it appears that if you working individually with the primary and the clone/snapshot.  But when you look at the details of the data in each, they are swapped.  Most recently, this has lead to data loss for us as we deleted what we thought was a stale (3 month old) clone/snapshot, but it was actually the primary dataset.  

### Describe how to reproduce the problem

We are using a ZFS volume to present a storage device to the OS on which we are building another filesystem, lustre ldiskfs in this case.  To back up this filesystem, we clone the ZFS volume, snapshot the clone, mount it, tar the data, then remove the snapshot and clone.  Occasionally, this process will get interrupted, leaving the clone/snapshot in place for some period of time.  We have found that what ZFS is presenting as the primary and the clone are actually swapped.  More details are provided in this thread. 

https://zfsonlinux.topicbox.com/groups/zfs-discuss/Tcb8a3ef663db0031-M5a79e71768b20b2389efc4a4 

I have not attempted to recreate this situation on a test system yet, but we've experienced it a couple of times in production now.  

### Include any warning/errors/backtraces from the system logs

None, which makes this a pretty nasty situation.  The data presented appears to be silently wrong.

**Comment by dvickernasa** [comment]
I have not attempted to recreate this situation on a test system yet, but I will see if I can reproduce this.  We have some monitor nodes in our lustre system that are running the same software stack so I have a good system to try to reproduce.

**Comment by dvickernasa** [comment]
I've created some scripts to setup a test to attempt to recreate this.  See the attached.  The `create.sh` script sets up a small volume (in an existing pool), builds an ext4 FS on it (simple and more accessible than the ldiskfs fs that this is occurring on in production), touches a file in the primary filesystem, makes a snapshot, clones that snapshot, and finally touches a new file in the original volume.   This leaves you with the primary volume containing two files and the clone with just one file, making it easy to see if one gets swapped with the other.  Note the ordering is slightly different than what I indicated in the description (snap then clone, not clone then snap).  Here is what this looks like on my test system after running `create.sh`.  

```
[root@hpfs-fsl-lmon0 test_vol_swap]# zfs list 
NAME                 USED  AVAIL     REFER  MOUNTPOINT
internal             145G  7.86T       24K  /internal
internal/mysql       131G  7.86T      105G  /internal/mysql
internal/mysql-OLD  13.5G  7.86T     13.5G  /internal/mysql-OLD
lmon0-0              242G   836G       38K  /lmon0-0
lmon0-0/mysql        240G   836G      145G  /lmon0-0/mysql
lmon0-0/test        1.03G   837G     41.5K  -
lmon0-0/test-clone  25.5K   836G       38K  -
[root@hpfs-fsl-lmon0 test_vol_swap]# ls -l /dev/zvol/lmon0-0/*
lrwxrwxrwx 1 root root  9 Sep 21 14:22 /dev/zvol/lmon0-0/test -> ../../zd0
lrwxrwxrwx 1 root root 10 Sep 21 14:22 /dev/zvol/lmon0-0/test-clone -> ../../zd16
[root@hpfs-fsl-lmon0 test_vol_swap]#  ls /mnt/test*
/mnt/test:
file1  file2  lost+found

/mnt/test-clone:
file1  lost+found
[root@hpfs-fsl-lmon0 test_vol_swap]# 

```

After I set this up I rebooted the machine several times, remounting both volumes and checking for a swap.  This did not recreate the issue so far - after all of the reboots `/mnt/test` contained the two files and `/mnt/test-clone` contained just the one file, as it should.   So no luck so far, but I will keep playing around with this.  At a minimum, these to at least provide the steps that are recreating this in production.  I'm happy to provide more info or run more tests as needed.  

[test_vol_swap.tgz](https://github.com/openzfs/zfs/files/12690982/test_vol_swap.tgz)

**Comment by GregorKopka** [comment]
> I have not attempted to recreate this situation on a test system yet, but I will see if I can reproduce this. We have some monitor nodes in our lustre system that are running the same software stack so I have a good system to try to reproduce.

You should be able to pull the sequence of commands from `zpool history` of the machine where you noticed the issue.

**Comment by dvickernasa** [comment]
I looked at that when it happened the first time and there was no indication from the history why they got swapped.  

I captured a full `zpool history` on the day this happened.  Here is the last part of that. 

```
[root@hpfs-fsl-mds1 ~]# grep meta-scratch zpool_history.2023.09.18.txt | tail
2023-06-18.03:33:05 zfs clone mds1-0/meta-scratch@snap mds1-0/meta-scratch-FORBACKUP
2023-06-18.21:21:53 zfs destroy mds1-0/meta-scratch-FORBACKUP
2023-06-18.21:22:24 zfs destroy mds1-0/meta-scratch@snap
2023-06-19.03:30:04 zfs snapshot mds1-0/meta-scratch@snap
2023-06-19.03:30:04 zfs clone mds1-0/meta-scratch@snap mds1-0/meta-scratch-FORBACKUP
2023-06-19.22:13:24 zfs destroy mds1-0/meta-scratch-FORBACKUP
2023-06-19.22:13:26 zfs destroy mds1-0/meta-scratch@snap
2023-06-20.03:11:09 zfs snapshot mds1-0/meta-scratch@snap
2023-06-20.03:11:09 zfs clone mds1-0/meta-scratch@snap mds1-0/meta-scratch-FORBACKUP
2023-09-18.14:47:22 zfs destroy mds1-0/meta-scratch-FORBACKUP
[root@hpfs-fsl-mds1 ~]# 
```

You can see when things got hung up on 6/20.  The destroy on 9/18 was done manually to get the process started again and this, in reality, deleted the primary dataset.  I will attach the full history if you'd like.

**Comment by dvickernasa** [comment]
Here is the full history.  
[zpool_history.2023.09.29.txt.gz](https://github.com/openzfs/zfs/files/12769260/zpool_history.2023.09.29.txt.gz)

**Comment by GregorKopka** [comment]
Do you by chance reference the exposed zvol through a partition in `/dev/disk/by-part*` instead of `/dev/zvol/...` ?

That _could_ result in creating the clone claiming the symlink of the original - which wouldn't be a problem per se, unless your upper filesystem restarts and reopens that device, now following the symlink to the clone instead of to the original...

Which could also explain why the automated destroy of the clone after the backup failed (and was not registered in `zpool history`): the device was in use by the upper filesystem.

So basically: could rebooting lustre _while the backup is running_ switch it from the original to the clone?

This would perfectly explain your results:
>clone destroy not happening
meta-scratch being at the state of 2023-06-20
meta-scratch-FORBACKUP containing the live state of your data

**Comment by dvickernasa** [comment]
No, we have been using just `/dev/zdX`.  That is what our lustre configuration file refers too. 

```
[root@hpfs-fsl-mds1 ~]# grep -e zd -e local /etc/ldev.conf
#local  foreign/-  label       [md|zfs:]device-path   [journal-path]/- [raidtab]
hpfs-fsl-mds1.fsl.jsc.nasa.gov  hpfs-fsl-mds0.fsl.jsc.nasa.gov  scratch-MDT0000  /dev/zd0
[root@hpfs-fsl-mds1 ~]# ls -l /dev/zd0
brw-rw---- 1 root disk 230, 0 Oct  1 09:14 /dev/zd0
[root@hpfs-fsl-mds1 ~]# 
```

All of the other locations (e.g. `/dev/zvol`, `/dev/disk/...`) are just sym links to the low level `/dev/zd0` devices.  And the problem seems to be that ZFS is getting these swapped under the hood.  This is from the 6/23/2022 post to the ZFS mailing list.  

```
[root@hpfs-fsl-mds1 ~]# ls -l /dev/zvol/mds1-0/
total 0
lrwxrwxrwx 1 root root  9 Jun 21 16:34 meta-scratch -> ../../zd0
lrwxrwxrwx 1 root root 10 Jun 21 16:34 meta-scratch-FORBACKUP -> ../../zd16
[root@hpfs-fsl-mds1 ~]# 
[root@hpfs-fsl-mds1 ~]# /lib/udev/zvol_id /dev/zd0
mds1-0/meta-scratch
[root@hpfs-fsl-mds1 ~]# /lib/udev/zvol_id /dev/zd16
mds1-0/meta-scratch-FORBACKUP
[root@hpfs-fsl-mds1 ~]# 
```

But if you read the rest of the posts from the mailing list, I could confirm by examining the data that `meta-scratch` and `meta-scratch-FORBACKUP` were really swapped.  Which means that `/dev/zd0` and `/dev/zd16' were swapped.  The latest incident from a couple weeks ago confirms this too.  When I destroyed `mds1-0/meta-scratch-FORBACKUP`, my primary data was lost.

**Comment by GregorKopka** [comment]
/dev/zd* are not guaranteed to be stable over pool exports/imports, which is something that seems to be doing with that pool on a regular basis (IIRC what I saw in the zpool history).

**Comment by GregorKopka** [comment]
No idea if there is actual documentation about `/dev/zd*` device nodes not being stable.

For me it was intuitive as the device nodes are created on-demand whenever a volume becomes available, same as the linux device mapper works (which has `/dev/dm-*` device nodes and `/dev/mapper/...` symlinks toward them to give a named handle to the mappings).

For ZFS volumes the `/dev/zvol/...` symlinks are stable, they always point toward the `/dev/zd*` device node of the block device backed by the named dataset.

**Comment by dvickernasa** [comment]
Fair enough.  We should switch to `/dev/zvol` for clairy if for no other reason because they make more sense to a person than `/dev/zdX`.   

However, that doesn't change the real problem, which is that ZFS is presenting these as swapped.  Last June, I verified the `/dev/zvol` links were pointing to what seemed like the right places but the data was swapped.  A couple of weeks ago, I didn't interact with the `/dev/zdX` devices at all.  I simply did a `zfs destroy mds1-0/meta-scratch-FORBACKUP` but in reality it destroyed the `mds1-0/meta-scratch` dataset.

**Comment by GregorKopka** [comment]
I read what you have experienced differently:
The `zfs destroy` _affected the right dataset_, the issue is that _your application wrote to the wrong one_ in the first place.

I have not researched if there is determinism in the order /dev/zd* device nodes are created on pool import, but my educated guess would be that reimporting a pool after creating or destroying volume datasets could certainly lead to all bets about _which volume ends up with what device node_ being off

**Comment by dvickernasa** [comment]
Interesting theory and, if you are correct about `/dev/zdX` not being stable, this is possible.

**Comment by GregorKopka** [comment]
> I simply did a `zfs destroy mds1-0/meta-scratch-FORBACKUP` but in reality it destroyed the `mds1-0/meta-scratch` dataset.

As a side note:
Having a pool checkpoint can save you from that (through with some downtime, as you'll have to export the pool to roll it back).
